node-init: Fixing pods restart on nodes running containerd on COS

fallard84 opened the pull request:
The startup script for the node-init was not properly working for nodes running containerd on COS for two reasons:
 1. crictl is not available in the PATH for root
 2. extracting the sandbox id using cat was returning an extra line

This fix is necessary for deployment on GKE using preemptible nodes. Otherwise, pods on nodes getting preempted might be created before Cilium is ready to manage them and they will remain that way unless they are properly stopped.

Fixes: #13167

```release-note
Fixing pods restart on nodes running containerd on COS
```

aanm commented on install/kubernetes/cilium/templates/cilium-nodeinit-daemonset.yaml:
```diff
@@ -220,7 +220,9 @@ spec:
                 # COS-beta (with containerd)
-                for f in `find /var/lib/cni/networks/ -type f ! -name lock ! -name last_reserved_ip.0`; do crictl stopp $(cat $f) || true; done
+                PATH=${PATH}:{{ .Values.cni.binPath }}
+                which crictl
```
How does it help running "which crictl"? Is it to make the script existing since it couldn't find such command and exit the node-init?

fallard84 replied:
I just found it useful debugging information. I can remove if you don't believe it is relevant to keep.

fallard84 replied:
Removed it as I am now calling crictl using the absolute path.

ti-mo commented on install/kubernetes/cilium/templates/cilium-nodeinit-daemonset.yaml:
```diff
@@ -220,7 +220,9 @@ spec:
               else
                 # COS-beta (with containerd)
-                for f in `find /var/lib/cni/networks/ -type f ! -name lock ! -name last_reserved_ip.0`; do crictl stopp $(cat $f) || true; done
+                PATH=${PATH}:{{ .Values.cni.binPath }}
```
I would call `crictl` by its absolute path instead of manipulating `PATH`. Why is `crictl` in `cni.binPath`? :thinking: Has this changed in recent versions of COS?

fallard84 replied:
I am not sure if that changed in previous version of COS. But I agree that it would be better to explicitely call `crictl` through its absolute path: `/home/kubernetes/bin/crictl`. I'll change that.

christarazi reviewed (changes requested):
Thanks for the PR. Change LGTM, Can you squash the second commit into the first?

ti-mo reviewed (approved):
Looks good, thanks for the fix!

kaworu reviewed (approved):
LGTM, thanks for the PR!

NOTE: no need for a full CI run.

aanm reviewed (changes requested):
@ti-mo @fallard84 we can't merge the PR as is at the risk of breaking older COS images. Can we either use `crictl` from PATH, and if not available, fallback to `/home/kubernetes/bin/crictl`? Thanks

Commit:
node-init: Fixing pods restart on COS containerd

The startup script for the node-init was not properly working for
nodes running containerd on COS for two reasons:
 1. crictl is not available in the PATH for root
 2. extracting the sandbox id using cat was returning an extra line

This fix is necessary for deployment on GKE using preemptible nodes.
Otherwise, pods on nodes getting preempted might be created before
Cilium is ready to manage them and they will remain that way unless
they are properly stopped.

Fixes: #13167

Signed-off-by: Francois Allard <francois@breathelife.com>

ti-mo commented on install/kubernetes/cilium/templates/cilium-nodeinit-daemonset.yaml:
```diff
@@ -220,7 +220,8 @@ spec:
               else
                 # COS-beta (with containerd)
-                for f in `find /var/lib/cni/networks/ -type f ! -name lock ! -name last_reserved_ip.0`; do crictl stopp $(cat $f) || true; done
+                which crictl || PATH=${PATH}:/home/kubernetes/bin
```
- FYI: `which` is not a shell built-in (at least not in Bash), the convention is to use `command -v` (see https://stackoverflow.com/questions/592620/how-can-i-check-if-a-program-exists-from-a-bash-script for explanation).
- Checking whether or not `crictl` exists here is arguably a bit pointless. There's no harm in having `/home/kubernetes/bin` in PATH even on older COS installs where this dir is not present.
- Please add a comment stating why this PATH extension is done. (e.g. required as of COS version x.y.z or any other useful info you might have) This helps with maintenance.

Thanks!

fallard84 replied:
Thanks, I wasn't aware that `which` wasn't recommended. I will make the requested changes.

joestringer reviewed (approved):
Thanks for the submission! No docs-structure impact, approving from that codeowner. The other feedback will still need to be addressed before merge though.
